aten/src/ATen/native/Math.h

```diff
@@ -978,76 +994,82 @@ static scalar_t _igam_helper_asymptotic_series(scalar_t a, scalar_t x, bool igam
       }
       ckterm = d[k][n]*etapow[n];
       ck += ckterm;
-      if (std::fabs(ckterm) < MACHEP * std::fabs(ck)) {
+      if (::fabs(ckterm) < MACHEP * ::fabs(ck)) {
         break;
       }
     }
     term = ck * afac;
-    absterm = std::fabs(term);
+    absterm = ::fabs(term);
     if (absterm > absoldterm) {
       break;
     }
     sum += term;
-    if (absterm < MACHEP * std::fabs(sum)) {
+    if (absterm < MACHEP * ::fabs(sum)) {
       break;
     }
     absoldterm = absterm;
     afac /= a;
   }
-  res += sgn * std::exp(-0.5 * a * eta * eta) * sum / std::sqrt(2 * c10::pi<float> * a);
+  res += (
+    sgn * ::exp(-HALF * a * eta * eta) * sum / ::sqrt(TWO * c10::pi<float> * a)
+  );
 
-  return res;
+  return static_cast<scalar_t>(res);
 }
 
-template <typename scalar_t>
-static scalar_t _igamc_helper_continued_fraction(scalar_t a, scalar_t x) {
+template <typename scalar_t, bool is_cuda=false>
+static C10_HOST_DEVICE scalar_t
+_igamc_helper_continued_fraction(scalar_t a, scalar_t x) {
   // Compute igamc using DLMF 8.9.2. [igam1]
+  using acc_t = at::acc_type<scalar_t, is_cuda>;
+  constexpr acc_t ZERO = acc_t{0.0};
+  constexpr acc_t ONE = acc_t{1.0};
+  constexpr acc_t TWO = acc_t{2.0};
+  const int MAXITER = 2000;
+  const acc_t MACHEP = std::is_same<scalar_t, double>::value ?
+    acc_t{1.11022302462515654042E-16} : acc_t{5.9604644775390625E-8};
+  const acc_t BIG = std::is_same<scalar_t, double>::value ?
+    acc_t{4.503599627370496e15} : acc_t{16777216.};
+  const acc_t BIGINV = std::is_same<scalar_t, double>::value ?
+    acc_t{2.22044604925031308085e-16} : acc_t{5.9604644775390625E-8};
   int i;
-  scalar_t ans, ax, c, yc, r, t, y, z;
-  scalar_t pk, pkm1, pkm2, qk, qkm1, qkm2;
-  int MAXITER = 2000;
-  static scalar_t MACHEP = std::is_same<scalar_t, double>::value ?
-    1.11022302462515654042E-16 : 5.9604644775390625E-8;
-  static scalar_t BIG = std::is_same<scalar_t,double>::value ?
-    4.503599627370496e15 : 16777216.;
-  static scalar_t BIGINV = std::is_same<scalar_t,double>::value ?
-    2.22044604925031308085e-16 : 5.9604644775390625E-8;
+  acc_t ans, ax, c, yc, r, t, y, z, pk, pkm1, pkm2, qk, qkm1, qkm2;
 
-  ax = _igam_helper_fac(a, x);
-  if (ax == 0.0) {
-    return 0.0;
+  ax = static_cast<acc_t>(_igam_helper_fac<scalar_t, is_cuda>(a, x));
```

Comment:
These functions should return `opmath_t` instead of truncating back to `scalar_t` precision.